File: RB101/exercises/medium_2/bubble_sort.rb
=begin
Bubble sort

within a loop
  until the switch counter == 0
    set a switch to 0
    set idx1 to 0
    set idx 2 to 0
    until idx2 == the size of the array
      if idx1 is greater than idx 2
        switch location of idx 1 and idx 2 within the array
        add 1 to switch variable
      end
      idx1 += 1
      idx2 += 1
    end
  end
  return the array
=end

def bubble_sort!(array)
  last_index = array.size - 1
  count = 0
  loop do
    switched = false
    1.upto(last_index) do |index|
      count += 1
      next if array[index - 1] <= array[index]
      array[index - 1], array[index] = array[index], array[index - 1]
      switched = true
      last_index = index
    end
    p count
    break unless switched
  end
  nil
end

p a = [4, 3, 2, 5, 1, 8, 9, 3, 44, 76, 22, 13, 67, 18, 17, 81].sort.reverse
p bubble_sort!(a)
p a

p array = %w(Sue Pete Alice Tyler Rachel Kim Bonnie)
p bubble_sort!(array)
p array

File: LS215/exercises/medium2/bubbleSort.js
/*
Questions:
Will the elements always be of the same data type? If not, how should we compare them?
Will there always be only one argument and wiil it be an array of elements? If not, what should we do?

Data:
input: array of elements of the same data type
output: the same array sorted

Algorithm:

sorted value to false;

while the sorted value is not true
  set sorted to true
  iterate over the indexs of elements up to second to last index
    if the first element is greater than the next element
      switch their positions
      set sorted to false
  
return the sorted array
*/

function bubbleSort(arr) {
  let sorted = false;

  while (!sorted) {
    sorted = true;

    for (let i = 0; i < arr.length - 1; i++) {
      if (arr[i] > arr[i + 1]) {
        swap(arr, i, i + 1)
        sorted = false;
      }
    }
  }

  return arr;
}

function swap(arr, idx1, idx2) {
  [arr[idx1], arr[idx2]] = [arr[idx2], arr[idx1]];
}

console.log(bubbleSort([])); // []
console.log(bubbleSort([2, 3, 1])); // [1, 2, 3]
console.log(bubbleSort(['z', 'x', 'a', 'c', 'b'])); // ['a', 'b', 'c', 'x', 'z']

const array1 = [5, 3];
bubbleSort(array1);
console.log(array1);    // [3, 5]

const array2 = [6, 2, 7, 1, 4];
bubbleSort(array2);
console.log(array2);    // [1, 2, 4, 6, 7]

const array3 = ['Sue', 'Pete', 'Alice', 'Tyler', 'Rachel', 'Kim', 'Bonnie'];
bubbleSort(array3);
console.log(array3);    // ["Alice", "Bonnie", "Kim", "Pete", "Rachel", "Sue", "Tyler"]